iceoryx_hoofs/source/posix_wrapper/named_pipe.cpp

```diff
@@ -196,49 +249,157 @@ cxx::expected<IpcChannelError> NamedPipe::timedSend(const std::string& message,
+cxx::expected<std::string, IpcChannelError> NamedPipe::timedReceive(const units::Duration& timeout) const noexcept
+{
+    if (!m_isInitialized)
+    {
+        return cxx::error<IpcChannelError>(IpcChannelError::NOT_INITIALIZED);
+    }
 
-        std::this_thread::sleep_for(std::chrono::nanoseconds(CYCLE_TIME.toNanoseconds()));
-        remainingTime = remainingTime - CYCLE_TIME;
-    } while (remainingTime.toNanoseconds() > 0U);
+    auto result = m_data->receiveSemaphore().timedWait(timeout);
+    cxx::Expects(!result.has_error());
 
+    if (*result == SemaphoreWaitState::NO_TIMEOUT)
+    {
+        auto message = m_data->messages.pop();
+        if (message.has_value())
+        {
+            cxx::Expects(!m_data->sendSemaphore().post().has_error());
+            return cxx::success<std::string>(message->c_str());
+        }
+        return cxx::error<IpcChannelError>(IpcChannelError::INTERNAL_LOGIC_ERROR);
+    }
     return cxx::error<IpcChannelError>(IpcChannelError::TIMEOUT);
 }
 
+NamedPipe::NamedPipeData::NamedPipeData(bool& isInitialized,
+                                        IpcChannelError& error,
+                                        const uint64_t maxMsgNumber) noexcept
+{
+    auto signalError = [&](const char* name) {
+        std::cerr << "Unable to create " << name << " semaphore for named pipe \"" << 'x' << "\"";
+        isInitialized = false;
+        error = IpcChannelError::INTERNAL_LOGIC_ERROR;
+    };
+
+    Semaphore::placementCreate(
+        &semaphores[SEND_SEMAPHORE], CreateUnnamedSharedMemorySemaphore, static_cast<unsigned int>(maxMsgNumber))
+        .or_else([&](auto) { signalError("send"); });
+
+    if (!isInitialized)
+    {
+        return;
+    }
+
+    Semaphore::placementCreate(&semaphores[RECEIVE_SEMAPHORE], CreateUnnamedSharedMemorySemaphore, 0U)
+        .or_else([&](auto) { signalError("receive"); });
+
+    if (!isInitialized)
+    {
+        return;
+    }
+
+    initializationGuard.store(VALID_DATA);
+}
+
+NamedPipe::NamedPipeData::~NamedPipeData() noexcept
+{
+    if (hasValidState())
+    {
+        sendSemaphore().~Semaphore();
+        receiveSemaphore().~Semaphore();
+    }
+}
+
+Semaphore& NamedPipe::NamedPipeData::sendSemaphore() noexcept
+{
+    return reinterpret_cast<Semaphore&>(semaphores[SEND_SEMAPHORE]);
+}
+
+Semaphore& NamedPipe::NamedPipeData::receiveSemaphore() noexcept
+{
+    return reinterpret_cast<Semaphore&>(semaphores[RECEIVE_SEMAPHORE]);
+}
+
+bool NamedPipe::NamedPipeData::waitForInitialization() const noexcept
+{
+    if (hasValidState())
+    {
+        return true;
+    }
+
+    units::Duration remainingTime = WAIT_FOR_INIT_TIMEOUT;
+
+    while (remainingTime.toNanoseconds() > 0U)
+    {
+        std::this_thread::sleep_for(std::chrono::nanoseconds(WAIT_FOR_INIT_SLEEP_TIME.toNanoseconds()));
+        remainingTime = remainingTime - WAIT_FOR_INIT_SLEEP_TIME;
+        if (hasValidState())
+        {
+            return true;
+        }
+    }
+
+    return false;
```

Comment:
theoretically we have a `DeadlineTimer` for stuff like this

Reply:
good hint!